src/libaktualizr/primary/sotauptaneclient.cc

```diff
@@ -873,116 +878,127 @@ result::UpdateStatus SotaUptaneClient::checkUpdatesOffline(const std::vector<Upt
 }
 
 result::Install SotaUptaneClient::uptaneInstall(const std::vector<Uptane::Target> &updates) {
-  result::Install result;
   const std::string &correlation_id = director_repo.getCorrelationId();
 
   // clear all old results first
   storage->clearInstallationResults();
 
-  // Recheck the Uptane metadata and make sure the requested updates are
-  // consistent with the stored metadata.
-  result::UpdateStatus update_status = checkUpdatesOffline(updates);
-  if (update_status != result::UpdateStatus::kUpdatesAvailable) {
-    if (update_status == result::UpdateStatus::kNoUpdatesAvailable) {
-      result.dev_report = {false, data::ResultCode::Numeric::kAlreadyProcessed, ""};
-    } else {
-      result.dev_report = {false, data::ResultCode::Numeric::kInternalError, ""};
+  // put most of the logic in a lambda so that we can take care of common
+  // post-operations
+  result::Install r;
+  std::string raw_report;
+  bool drop_targets = false;
+
+  std::tie(r, raw_report, drop_targets) = [this, &updates,
+                                           &correlation_id]() -> std::tuple<result::Install, std::string, bool> {
```

Comment:
This is a great idea and should really make things easier to deal with. Took me a minute to follow it, though. I think part of it is that the lambda is rather large. Would it perhaps make it easier to just make it a proper function? Or would that be unnecessary? I realize most of this function *is* the lambda now, so maybe that wouldn't really help things...

Reply:
I tend to agree actually but my bad excuse is that I don't really know how to name this function, it's really doing the bulk of the work...

The `do { ... break; } while(0);` idiom was considered but a return is in a nested loop, so it would not work.

Another note is that `return std::make_tuple(...)` can be changed to `return {...}` from c++14 http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/n4387.html. The first version was written in this style but it tripped on xenial.